pkg/referrerstore/oras/authprovider/azure/azureworkloadidentity.go

```diff
@@ -0,0 +1,158 @@
+func init() {
+	provider.Register(azureWIAuthProviderName, &AzureWIProviderFactory{})
+}
+
+// Create returns an AzureWIAuthProvider
+func (s *AzureWIProviderFactory) Create(authProviderConfig provider.AuthProviderConfig) (provider.AuthProvider, error) {
+	conf := azureWIAuthProviderConf{}
+	authProviderConfigBytes, err := json.Marshal(authProviderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(authProviderConfigBytes, &conf); err != nil {
+		return nil, fmt.Errorf("failed to parse auth provider configuration: %v", err)
+	}
+
+	return &azureWIAuthProvider{}, nil
+}
+
+// Enabled always returns true since there are no fields to verify
+func (d *azureWIAuthProvider) Enabled() bool {
+	return true
+}
+
+// Provide returns the credentials for a specificed artifact.
+// Uses Azure Workload Identity to retrieve an AAD access token which can be
+// exchanged for a valid ACR refresh token for login.
+func (d *azureWIAuthProvider) Provide(artifact string) (provider.AuthConfig, error) {
+	tenantID := os.Getenv("AZURE_TENANT_ID")
+
+	// parse the artifact reference string to extract the registry host name
+	artifactHostName, err := provider.GetRegistryHostName(artifact)
+	if err != nil {
+		return provider.AuthConfig{}, err
+	}
+
+	// retrieve an AAD Access token
+	aadToken, err := getAADAccessToken(tenantID, AADResource)
+	if err != nil {
+		return provider.AuthConfig{}, err
+	}
+
+	// send a challenge to the login server
+	directive, err := receiveChallengeFromLoginServer(artifactHostName, "https")
+	if err != nil {
+		return provider.AuthConfig{}, err
+	}
+
+	// use challenge directive and AAD token to exchange for a registry token
+	refreshToken, err := performTokenExchange(artifactHostName, directive, tenantID, aadToken)
+	if err != nil {
+		return provider.AuthConfig{}, err
+	}
+
+	authConfig := provider.AuthConfig{
+		Username: dockerTokenLoginUsernameGUID,
+		Password: refreshToken,
+		Provider: d,
+	}
+	return authConfig, nil
+}
+
+// Source: https://github.com/Azure/azure-workload-identity/blob/d126293e3c7c669378b225ad1b1f29cf6af4e56d/examples/msal-go/token_credential.go#L25
+func getAADAccessToken(tenantID, resource string) (string, error) {
+	// Azure AD Workload Identity webhook will inject the following env vars:
+	// 	AZURE_CLIENT_ID with the clientID set in the service account annotation
+	// 	AZURE_TENANT_ID with the tenantID set in the service account annotation. If not defined, then
+	// 	the tenantID provided via azure-wi-webhook-config for the webhook will be used.
+	// 	AZURE_FEDERATED_TOKEN_FILE is the service account token path
+	// 	AZURE_AUTHORITY_HOST is the AAD authority hostname
+	clientID := os.Getenv("AZURE_CLIENT_ID")
+	tokenFilePath := os.Getenv("AZURE_FEDERATED_TOKEN_FILE")
+	authorityHost := os.Getenv("AZURE_AUTHORITY_HOST")
+
+	// read the service account token from the filesystem
+	signedAssertion, err := readJWTFromFS(tokenFilePath)
```

Comment:
How often these envs change? For example how often the token file changes? If they are injected when the pod is provisioned, do you need to read this file for every request?

Reply:
I believe the token file is injected into the pod on provision. I'll need to look into it to confirm. If that's the case, then we can handle the AAD token operation on pod startup in auth provider registration process rather than on each Provide call.

Reply:
Is this still pending?

Reply:
I have implemented the change. Now the initial AAD token exchange happens on pod startup and not every time Provide is called. I also added an ExpiresOn field for the credentials returned to account for the 24 hour AAD token life span. I also added logic to check for token expiry and call a new AAD token exchange as necessary.